### PR TITLE
[GSF tracking] Add full Sylvester criterion for positive definite check

### PR DESCRIPTION
**Pull request by swagata87**
#### PR description:
This PR proposes to add a more strict check of positive definiteness in order to avoid GSF crashes observed recently at HLT as discussed in this issue https://github.com/cms-sw/cmssw/issues/39570.

#### PR validation:

`runTheMatrix.py -l 12434.0` ran fine.

In order to check if this PR would affect HLT electron efficiency, I reran HLT on EGamma RAW data, before and after this PR, and checked trigger decisions for 1000 events. HLT menu `/online/collisions/2022/2e34/v1.4/HLT/V2` was run on `/eos/cms/tier0/store/data/Run2022E/EGamma/RAW/v1/000/359/871/00000/fcf81596-54ed-45d3-9dee-32dec392f246.root`. Trigger result was same, before and after this PR. For example:
```
HLT name | nEventPassed before PR | nEventPassed after PR
HLT_Ele32_WPTight_Gsf_v17 | 376 | 376
HLT_Ele23_Ele12_CaloIdL_TrackIdL_IsoVL_DZ_v21 | 30 | 30
```

The PR test can be done preferably by enabling profiling to see if there is any extra CPU cost for the computation of determinants of sub-matrices.

If this fix is accepted, then a backport to data-taking release cycle would be nice.

### CODE REVIEW

**Comment by slava77** [comment]
type tracking,egamma

File: TrackingTools/GsfTracking/src/GsfMultiStateUpdator.cc [hunk 1]
```diff
@@ -32,7 +32,13 @@ TrajectoryStateOnSurface GsfMultiStateUpdator::update(const TrajectoryStateOnSur
...
 
     double det = 0;
+    double det22 = 0;
+    double det33 = 0;
+    double det44 = 0;
     if (updatedTSOS.isValid() && updatedTSOS.localError().valid() && updatedTSOS.localError().posDef() &&
+        (updatedTSOS.curvilinearError().matrix().Sub<AlgebraicSymMatrix22>(0, 0).Det(det22) && det22 > 0) &&
+        (updatedTSOS.curvilinearError().matrix().Sub<AlgebraicSymMatrix33>(0, 0).Det(det33) && det33 > 0) &&
+        (updatedTSOS.curvilinearError().matrix().Sub<AlgebraicSymMatrix44>(0, 0).Det(det44) && det44 > 0) &&
         (updatedTSOS.curvilinearError().matrix().Det2(det)) && det > 0) {
```
Comment (fwyzard) [inline TrackingTools/GsfTracking/src/GsfMultiStateUpdator.cc:42]: You can simplify this to
```suggestion
    if (double det;
        updatedTSOS.isValid() && updatedTSOS.localError().valid() && updatedTSOS.localError().posDef() &&
        (det = 0., updatedTSOS.curvilinearError().matrix().Sub<AlgebraicSymMatrix22>(0, 0).Det(det) && det > 0) &&
        (det = 0., updatedTSOS.curvilinearError().matrix().Sub<AlgebraicSymMatrix33>(0, 0).Det(det) && det > 0) &&
        (det = 0., updatedTSOS.curvilinearError().matrix().Sub<AlgebraicSymMatrix44>(0, 0).Det(det) && det > 0) &&
        (det = 0., updatedTSOS.curvilinearError().matrix().Det2(det) && det > 0))
    {
```

Comment (swagata87) [reply TrackingTools/GsfTracking/src/GsfMultiStateUpdator.cc]: thanks Andrea, now implemented

**Comment by mandrenguyen** [comment]
please test

**Comment by swagata87** [comment]
A timing measurement for HLT was done in `cms-hlt-gpu.cern.ch` in `CMSSW_12_4_9` comparing `Base` vs `Base+this PR`.
(#jobs, #threads, #streams)=(8, 8, 0), and nEvents=10000.
The results are given below.


**Base**



`DST_Run3_PFScoutingPixelTracking_v` : 167.6 ms


**Base+this PR**



`DST_Run3_PFScoutingPixelTracking_v` : 170.4 ms

FYI: @cms-sw/hlt-l2

**Comment by mandrenguyen** [comment]
enable profiling
(apologies, missed this request in the PR description)

**Comment by mandrenguyen** [comment]
please test

**Comment by mandrenguyen** [comment]
+1
it seems there is a small increase in time to
`CkfTrackCandidateMakerBase::produceBase `
but that the total change of CPU is within the margin of error for the profiling.

**Comment by perrotta** [comment]
+1

**Pull request merged**